symengine/polys/basic_conversions.h

```diff
@@ -0,0 +1,227 @@
+            dict = P::container_from_dict(gen, {{0, i}});
+    }
+
+    void bvisit(const Basic &x)
+    {
+        RCP<const Basic> genpow = one, genbase = gen, powr;
+        if (is_a<const Pow>(*gen)) {
+            genpow = rcp_static_cast<const Pow>(gen)->get_exp();
+            genbase = rcp_static_cast<const Pow>(gen)->get_base();
+        }
+
+        if (eq(*genbase, x)) {
+            powr = div(one, genpow);
+            if (is_a<const Integer>(*powr)) {
+                int i = rcp_static_cast<const Integer>(powr)->as_int();
+                if (i > 0) {
+                    dict = P::container_from_dict(gen, {{i, 1}});
+                    return;
+                }
+            }
+        }
+
+        d_add_term(0, x);
+    }
+};
+
+template <typename Poly>
+class BasicToUIntPoly : public BasicToUPolyBase<Poly, BasicToUIntPoly<Poly>>
+{
+public:
+    using BasicToUPolyBase<Poly, BasicToUIntPoly>::bvisit;
+    using BasicToUPolyBase<Poly, BasicToUIntPoly>::apply;
+
+    void bvisit(const Rational &x)
+    {
+        throw std::runtime_error("Non-integer found");
+    }
+
+    void d_add_term(unsigned int pow, const Basic &x)
+    {
+        if (is_a<const Integer>(x))
+            this->dict = Poly::container_from_dict(
+                this->gen, {{pow, static_cast<const Integer &>(x).i}});
+        else
+            throw std::runtime_error("Non-integer found");
+    }
+};
+
+class BasicToUExprPoly : public BasicToUPolyBase<UExprPoly, BasicToUExprPoly>
+{
+public:
+    using BasicToUPolyBase<UExprPoly, BasicToUExprPoly>::bvisit;
+    using BasicToUPolyBase<UExprPoly, BasicToUExprPoly>::apply;
+
+    void bvisit(const Rational &x)
+    {
+        if (not x.is_zero())
+            dict = UExprDict(x.rcp_from_this());
+    }
+
+    void d_add_term(unsigned int pow, const Basic &x)
+    {
+        dict = UExprDict({{pow, x.rcp_from_this()}});
+    }
+};
+
+template <typename T, typename P>
+enable_if_t<std::is_same<T, UExprDict>::value, T>
+_b_to_upoly(const RCP<const Basic> &basic, const RCP<const Basic> &gen)
+{
+    BasicToUExprPoly v;
+    return v.apply(*basic, gen);
+}
+
+template <typename T, typename P>
+enable_if_t<std::is_same<T, UIntDict>::value
+#ifdef HAVE_SYMENGINE_PIRANHA
+                or std::is_same<T, pintpoly>::value
+#endif
+#ifdef HAVE_SYMENGINE_FLINT
+                or std::is_same<T, fzp_t>::value
+#endif
+            ,
+            T>
+_b_to_upoly(const RCP<const Basic> &basic, const RCP<const Basic> &gen)
+{
+    BasicToUIntPoly<P> v;
+    return v.apply(*basic, gen);
+}
+
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic,
+                        const RCP<const Basic> &gen, bool ex = false)
+{
+    RCP<const Basic> exp = basic;
+    if (ex)
+        exp = expand(basic);
+    return P::from_container(
+        gen, _b_to_upoly<typename P::container_type, P>(exp, gen));
+}
+
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex = false)
```

Comment:
Can you copy this definition to the top of the file and keep the implementation here? It'll be easier to read about what the header is providing.